### PR TITLE
Switch to wiping user records, instead of deleting them

### LINKED ISSUE #620: Mutations for a user to deactivate & delete themselves
Deactivation:
- Back-end: set their active column to false. Log them out and delete their session. 
- If they try to log back in say this will re-activate your account, first asking them if they are sure they want to do that. 
- What should we show for posts and messages with deactivated users? Leave content for now we think... make sure people know it will be left up
- Show them in membership directory? No

Delete account:
- We don't delete the posts, we just delete the content of the post and change it to say Deleted Post, but leave comments on the post by other people. And change comments to say Deleted Comment. We will likely need a user that is called Deleted User that we point deleted posts and comments too as the creator/owner.
- What about upvotes? Could change to point to the Deleted User row in the users table, so we still keep the count the same...?
- Do we delete the groups they created? If other moderators we leave up as normal? If they were the only moderator in the group we pause/deactivate the group. Don't delete it but no more activity can happen in it. (This will also cause it to disappear for members of the group.) Do we need to send them a notification saying the group was deactivated?

### PR DESCRIPTION
**Pull request by thomasgwatson**
Closes #620

### CODE REVIEW

File: api/models/User.js [hunk 1]
```diff
@@ -183,30 +182,28 @@ module.exports = bookshelf.Model.extend(merge({
       - delete their user record!
     */
 
+    // look up their uploads and delete
```
Comment (thomasgwatson) [inline api/models/User.js:185]: Somewhat blocked on what to do here. I am assuming there is something happening outside of the node code that is putting files into a CDN, after they are inserted into s3. And our db records seem to mainly reference the CDN urls. 

So to delete, I have to reconstruct s3 keys from a file's CDN url? Is that possible? And then I have to delete the s3 object and also invalidate the CDN url record?

**Comment by tibetsprague** [comment]
We just upload stuff to S3 and then cloud front sits in front of S3. The URLs for S3 and cloud front are the same, just with a different host. See getFinalUrl in lib/uploader/storage.js
So you can just find the same path in S3 and delete it.

File: api/models/User.js [hunk 2]
```diff
@@ -170,43 +171,50 @@ module.exports = bookshelf.Model.extend(merge({
...
-    UPDATE comments SET deactivated_by_id = ${deletedUserId} WHERE deactivated_by_id = ${this.id};
     UPDATE comments SET text = 'Comment by deleted user' WHERE user_id = ${this.id};
 
-    UPDATE follows set added_by_id = ${deletedUserId} WHERE added_by_id = ${this.id};
-    UPDATE groups_tags set user_id = ${deletedUserId} WHERE user_id = ${this.id};
     DELETE FROM thanks WHERE comment_id in (select id from comments WHERE user_id = ${this.id});
     DELETE FROM thanks WHERE thanked_by_id = ${this.id};
     DELETE FROM notifications WHERE activity_id in (select id from activities WHERE reader_id = ${this.id});
     DELETE FROM notifications WHERE activity_id in (select id from activities WHERE actor_id = ${this.id});
+    DELETE FROM notifications WHERE activity_id in (select id from activities WHERE parent_comment_id in (select id from comments where user_id = ${this.id}));
```
Comment (tibetsprague) [inline api/models/User.js:210]: so now that im thinking about this, won't this delete notifications for other people that have comments on this person's comments? I think we need to check that the notification or the activity also is for this particular user? im a bit confused...

File: api/models/User.js [hunk 3]
```diff
@@ -221,7 +229,31 @@ module.exports = bookshelf.Model.extend(merge({
     DELETE FROM posts_tags WHERE post_id in (select id from posts WHERE user_id = ${this.id});
     DELETE FROM groups_posts WHERE post_id in (select id from posts WHERE user_id = ${this.id});
     DELETE FROM votes WHERE user_id = ${this.id};
-    DELETE FROM users WHERE id = ${this.id};
+
+    UPDATE users SET 
+    active = false,
```
Comment (tibetsprague) [inline api/models/User.js:234]: TBD

File: api/models/User.js [hunk 4]
```diff
@@ -221,7 +229,31 @@ module.exports = bookshelf.Model.extend(merge({
...
+    bio = NULL, 
+    banner_url = NULL,
+    location = NULL,
+    url = NULL,
+    tagline = NULL,
+    stripe_account_id = NULL,
+    location_id = NULL,
+    contact_email = NULL,
+    contact_phone = NULL,
+    email = '${uuid.v4()}@hylo.com',
```
Comment (tibetsprague) [inline api/models/User.js:247]: still thinking about this one too... I guess this looks best

Comment (thomasgwatson) [reply api/models/User.js:247]: Yeah, just needed a unique value

File: api/models/User.js (hunk 2 above)
Comment (thomasgwatson) [reply api/models/User.js:210]: I think this was added to ensure its counterpart of

`DELETE from activities WHERE parent_comment_id in (select id from comments where user_id = ${this.id});` would work without hitting a constraint.

However, I wonder now whether this `DELETE from` is even necessary and whether they can both be removed?

File: api/models/User.js (hunk 3 above)
Comment (thomasgwatson) [reply api/models/User.js:234]: What are our options here if we flip back to active, considering we would want to turn off notifications? When sending notifications, check the user name and if it matches "Deleted User", filter it out? Are there other things we'd want to skip doing?

Alternatively, if its just that we want posts and comments from deleted users to still show up, do we just change the current queries for those to now filter out non-active users?? What non-active users do we actually not want to see posts or comments for?

File: api/models/User.js (hunk 2 above)
Comment (tibetsprague) [reply api/models/User.js:210]: I think we should talk this through

File: api/models/User.js (hunk 3 above)
Comment (tibetsprague) [reply api/models/User.js:234]: yes it seems like the whole point is to have the posts and comments still show up. but im not sure we want them to show up for people who mark themselves as deactivated. again, lets talk it through

**Review by tibetsprague — APPROVED** [review]
this seems ready to test again!

**Pull request merged**